tracer/test/Datadog.Trace.Tests/SpanContextPropagatorTests.cs

```diff
@@ -3,34 +3,322 @@
+            // ignore the extracted sampling priority if it is not a valid integer
+
+            var headers = SetupMockHeadersCollection();
+
+            // replace SamplingPriority setup
+            headers.Setup(h => h.GetValues(HttpHeaderNames.SamplingPriority)).Returns(new[] { samplingPriority });
+
+            object result = SpanContextPropagator.Instance.Extract(headers.Object);
+
+            result.Should()
+                  .BeEquivalentTo(
+                       new
+                       {
+                           TraceId,
+                           SpanId,
+                           Origin,
+                           SamplingPriority = default(SamplingPriority?)
+                       });
+        }
+
+        private static Mock<IHeadersCollection> SetupMockHeadersCollection()
+        {
+            return SetupMockHeadersCollection(TraceId.ToString(), SpanId.ToString());
+        }
+
+        private static Mock<IHeadersCollection> SetupMockHeadersCollection(string traceId, string spanId)
+        {
+            var headers = new Mock<IHeadersCollection>(MockBehavior.Strict);
+            headers.Setup(h => h.GetValues(HttpHeaderNames.TraceId)).Returns(new[] { traceId });
+            headers.Setup(h => h.GetValues(HttpHeaderNames.ParentId)).Returns(new[] { spanId });
+            headers.Setup(h => h.GetValues(HttpHeaderNames.SamplingPriority)).Returns(new[] { ((int)SamplingPriority).ToString() });
+            headers.Setup(h => h.GetValues(HttpHeaderNames.Origin)).Returns(new[] { Origin });
+            return headers;
+        }
+
+        private static Mock<IReadOnlyDictionary<string, string>> SetupMockReadOnlyDictionary()
+        {
+            var headers = new Mock<IReadOnlyDictionary<string, string>>();
+
+            var traceId = TraceId.ToString();
+            headers.Setup(h => h.TryGetValue(HttpHeaderNames.TraceId, out traceId)).Returns(true);
+
+            var spanId = SpanId.ToString();
+            headers.Setup(h => h.TryGetValue(HttpHeaderNames.ParentId, out spanId)).Returns(true);
+
+            var samplingPriority = ((int)SamplingPriority).ToString();
+            headers.Setup(h => h.TryGetValue(HttpHeaderNames.SamplingPriority, out samplingPriority)).Returns(true);
+
+            var origin = Origin;
+            headers.Setup(h => h.TryGetValue(HttpHeaderNames.Origin, out origin)).Returns(true);
+
+            return headers;
+        }
+
+        private static void VerifySetCalls(Mock<IHeadersCollection> headers)
+        {
+            var once = Times.Once();
+
+            headers.Verify(h => h.Set(HttpHeaderNames.TraceId, TraceId.ToString()), once);
+            headers.Verify(h => h.Set(HttpHeaderNames.ParentId, SpanId.ToString()), once);
+            headers.Verify(h => h.Set(HttpHeaderNames.SamplingPriority, ((int)SamplingPriority).ToString()), once);
+            headers.Verify(h => h.Set(HttpHeaderNames.Origin, Origin), once);
+            headers.VerifyNoOtherCalls();
```

Comment:
It's maybe just me, but I'm really not a fan of tests that use mocks to monitor the exact calls were made, as it makes them super-brittle. I'd prefer a stub `IHeadersCollection` stub that verifies the final state. But maybe that's just my "app-developer" viewpoint rather than "performance" viewpoint, as I guess we _would_ want these tests to fail if we were setting the value twice (for example)? 🤔 Thinking out loud here 😅

Reply:
> we would want these tests to fail if we were setting the value twice (for example)

Yeah, I was thinking we want the behavior of `SpanContextPropagator` (and what it does with the `IHeadersCollection`) to be well-defined. If we call `Set()` on the same header twice, it's probably a bug.